### PR TITLE
Added quotas API write methods

### PR DESCRIPTION
**Pull request by turegjorup**
Thank you for your advice in #654. 

This is my first attempt at implementing create/update/delete for the `quotas` endpoint. There are a couple of things I'm unsure of, so I have marked the PR as WIP:

* I haven't found any validation rules for quotas besides the model validation - is this correct?
* I have added a couple of tests, but I'm unsure of what the edge cases would be, so this could definitely be expanded upon.

Code checks and unit tests _should_ pass.

### CODE REVIEW

File: doc/api/resources/quotas.rst [hunk 1]
```diff
@@ -106,6 +106,134 @@ Endpoints
...
+        "subevent": null
+      }
+
+   :param organizer: The ``slug`` field of the organizer to create a voucher for
+   :param event: The ``slug`` field of the event to create a voucher for
+   :statuscode 201: no error
+   :statuscode 400: The voucher could not be created due to invalid submitted data.
+   :statuscode 401: Authentication failure
+   :statuscode 403: The requested organizer/event does not exist **or** you have no permission to create this resource.
+   :statuscode 409: The server was unable to acquire a lock and could not process your request. You can try again after a short waiting period.
```
Comment (raphaelm) [inline doc/api/resources/quotas.rst]: 409 can not happen for this operation, so we can drop it.

File: doc/api/resources/quotas.rst [hunk 2]
```diff
@@ -106,6 +106,134 @@ Endpoints
...
+        "size": 200,
+        "items": [1, 2],
+        "variations": [1, 4, 5, 7],
+        "subevent": null
+      }
+
+   :param organizer: The ``slug`` field of the organizer to create a voucher for
+   :param event: The ``slug`` field of the event to create a voucher for
+   :statuscode 201: no error
+   :statuscode 400: The voucher could not be created due to invalid submitted data.
```
Comment (raphaelm) [inline doc/api/resources/quotas.rst]: voucher → quota

File: doc/api/resources/quotas.rst [hunk 3]
```diff
@@ -106,6 +106,134 @@ Endpoints
...
+      }
+
+   :param organizer: The ``slug`` field of the organizer to modify
+   :param event: The ``slug`` field of the event to modify
+   :param id: The ``id`` field of the quota rule to modify
+   :statuscode 200: no error
+   :statuscode 400: The quota could not be modified due to invalid submitted data
+   :statuscode 401: Authentication failure
+   :statuscode 403: The requested organizer/event does not exist **or** you have no permission to change this resource.
+   :statuscode 409: The server was unable to acquire a lock and could not process your request. You can try again after a short waiting period.
```
Comment (raphaelm) [inline doc/api/resources/quotas.rst]: Same here, 409 won't happen

File: src/pretix/api/views/item.py [hunk 4]
```diff
@@ -77,18 +78,58 @@ class Meta:
...
     ordering = ('id',)
     permission = 'can_change_items'
+    write_permission = 'can_change_items'
 
     def get_queryset(self):
         return self.request.event.quotas.all()
 
+    def create(self, request, *args, **kwargs):
+        with request.event.lock():
+            return super().create(request, *args, **kwargs)
```
Comment (raphaelm) [inline src/pretix/api/views/item.py]: We don't need the lock here as we do with vouchers and can therefore drop this method completely.

File: src/pretix/api/views/item.py [hunk 5]
```diff
@@ -77,18 +78,58 @@ class Meta:
...
+        )
+
+    def get_serializer_context(self):
+        ctx = super().get_serializer_context()
+        ctx['event'] = self.request.event
+        return ctx
+
+    def update(self, request, *args, **kwargs):
+        with request.event.lock():
+            return super().update(request, *args, **kwargs)
```
Comment (raphaelm) [inline src/pretix/api/views/item.py]: We don't need the lock here as we do with vouchers and can therefore drop this method completely.

File: src/pretix/api/views/item.py [hunk 6]
```diff
@@ -77,18 +78,58 @@ class Meta:
...
+
+    def perform_update(self, serializer):
+        serializer.save(event=self.request.event)
+        serializer.instance.log_action(
+            'pretix.event.quota.changed',
+            user=self.request.user,
+            api_token=(self.request.auth if isinstance(self.request.auth, TeamAPIToken) else None),
+            data=self.request.data
+        )
```
Comment (raphaelm) [inline src/pretix/api/views/item.py]: In the form view, we do a quite elaborate logging based on if the associated subevent changed. The reason is that we do not only want the log to show up in the quota detail view but in the subevent's log as well. I just saw that we only do this during updating, not during deletion or creation (which is kinda inconsistent), but we should at least keep API and form behaviour consistent. Reference: https://github.com/pretix/pretix/blob/master/src/pretix/control/views/item.py#L685

Also, we should call ``serializer.instance.rebuild_cache()`` after saving here.

**Comment by raphaelm** [comment]
Hi,

thank you very much for the high-quality PR! :sparkles:  I've added a few minor comments with pretix-specific things you couldn't have known and I have some answers to your questions.

> I haven't found any validation rules for quotas besides the model validation - is this correct?

Heh. Yes and no. This is why I suggested this as a starter, because it has no real explicit validation rules. The form does however have some *implicit* validation baked into its structure that is really non-obvious. As I see it, they are:

* It must be forbidden to add a variation to the ``variations`` list that does not belong to an item contained in the ``items`` list
* It must be forbidden to remove an item from the ``items`` list if one of its ``variations`` is still in the ``variations`` list
* It must be forbidden to add an item to the ``items`` ist if it *has* variations but none of its variations is in the ``variations`` list
* (most important one) It must be forbidden to link an item or variation that belongs to a different event – this would allow to hijack other organizer's ticket shops otherwise
* It must be forbidden to link a subevent that belongs to a different event

> I have added a couple of tests, but I'm unsure of what the edge cases would be, so this could definitely be expanded upon.

Your tests are fine, but with the additional validation constraints I outlined above it should be clear how to come up with some interesting edge cases :wink: 

>Code checks and unit tests should pass.

I'm really unsure why Travis does not build this PR :confused: 
If it doesn't start to do so, I'll check manually as soon as this is done with.

Let me know if you have any more questions! :)

**Comment by raphaelm** [comment]
Also, forgot that to mention, you should add the new URLs to the test parameter set in ``tests/api/test_permissions.py`` to make sure the permission requirements are tested. You can look at the voucher lines: The fourth parameter is the expected return code for "permission granted" which is 400/404 in this case since the test will send invalid data (does not matter for checking the permissions).

**Comment by turegjorup** [comment]
Thank you for your feedback @raphaelm. I have attempted to address most of it:
* `tests/api/test_permissions.py` updated with new URLs
* validation rules for items/variations added

I still need to address logging and subevents

File: src/pretix/base/models/items.py [hunk 7]
```diff
@@ -914,3 +914,11 @@ def clean_items(event, items, variations):
...
                 if bool(selected_vars) is False:
                     raise ValidationError(_('One or more items has variations but none of these are in the variations list'))
+
+    @staticmethod
+    def clean_subevent(event, subevent):
+        if event.has_subevents:
+            if not subevent:
+                raise ValidationError(_('Subevent cannot be null for event series'))
+            if event != subevent.event:
+                raise ValidationError(_('The subevent does not belong to this event'))
```
Comment (raphaelm) [inline src/pretix/base/models/items.py]: At the same time, we could check that subevent *is* null if event.has_subevents is False :)

**Comment by turegjorup** [comment]
Hi @raphaelm - I believe the last commits fixes the issues you have mentioned. If not please advise where further improvements are needed :-)

The only thing I'm not completely satisfied with is the fixtures in the tests. I think we should consider how best to structure them going forward. So far, I have used names like "Event2" and "Event3" but this doesn't really communicate either the content or the intended usage.

I think we should consider these questions:
* Should we name fixtures after content, e.g. "event_with_subevent", "item_with_variations", etc?
* Can/shall fixtures be shared among tests, e.g. `conftest.py` or should we work towards more dedicated fixtures for individual tests? The former will give less code duplication but probably a more complex fixtures structure. The alter will cause us to duplicate fixtures but might enable more focused/less complex fixtures for the individual tests.

Tests pass with a few exceptions:
```   
1729 passed
1 xpassed
1 failed
- tests/control/test_settings.py:301 MailSettingPreviewTest.test_localised_payment_info
1 skipped
3 rerun
```
Fail:
```
E       AssertionError: assert 'Please transfer money to this bank account: 9999-9999-9999-9999' != 'Please transfer money to this bank account: 9999-9999-9999-9999'

tests/control/test_settings.py:313
```
xpassed:
```
tests/api/test_orders.py
```
In both cases it seems unrelated to the PR so I'm unsure how to address this?

**Comment by raphaelm** [comment]
Hey! :) Thanks, I'll look into it more deeply as soon as I get around to.

> The only thing I'm not completely satisfied with is the fixtures in the tests. I think we should consider how best to structure them going forward. So far, I have used names like "Event2" and "Event3" but this doesn't really communicate either the content or the intended usage.

I can relate to that, this absolutely something I'm not satisfied with in the pretix codebase. This is probably something to do a large refactoring about but I'm also really unsure how the final design should look. We could go for more unified fixtures in conftest files or for using [FactoryBoy](https://factoryboy.readthedocs.io/en/latest/) instead of py.test fixtures in places where we need very individual fixtures. 

> The former will give less code duplication but probably a more complex fixtures structure.

This is exactly the trade-off I'm struggling to make. I don't know, yet, so for now, I don't care about it too much for the scope of this PR.

> Should we name fixtures after content, e.g. "event_with_subevent", "item_with_variations", etc?

Yeah, that's probably better. I don't like pytests fixture system here as it leads to long parameter and therefore long variable names which makes the tests harder to write, but as long as it's easier to read, then… ;) If you only need a very simple fixture once or twice I'd probably drop the fixture and inline it to the test.

> Fail: tests/control/test_settings.py:313

This is just because you did not compile translation files (``make localecompile``)

> xpassed: tests/api/test_orders.py

This is my fault and can be ignored for now :wink:

**Comment by turegjorup** [comment]
> This is exactly the trade-off I'm struggling to make. I don't know, yet, so for now, I don't care about it too much for the scope of this PR.

OK - I'll leave it as it is for now. Looking forward to further feedback 😃

File: doc/api/resources/quotas.rst [hunk 8]
```diff
@@ -106,6 +106,132 @@ Endpoints
...
 
+.. http:post:: /api/v1/organizers/(organizer)/events/(event)/quotas/
+
+   Creates a new quota
+
+   **Example request**:
+
+   .. sourcecode:: http
+
+      POST /api/v1/organizers/bigevents/events/sampleconf/quotas/1/ HTTP/1.1
```
Comment (raphaelm) [inline doc/api/resources/quotas.rst]: this is only to ``/quotas/``, not ``/quotas/1/``, I guess :wink:

File: src/pretix/api/views/item.py [hunk 9]
```diff
@@ -77,18 +78,71 @@ class Meta:
...
+        if serializer.instance.subevent:
+            serializer.instance.subevent.log_action(
+                'pretix.subevent.quota.changed',
+                user=self.request.user,
+                api_token=(self.request.auth if isinstance(self.request.auth, TeamAPIToken) else None),
+                data=self.request.data
+            )
```
Comment (raphaelm) [inline src/pretix/api/views/item.py]: This still misses the subtle case in which *changing* the subevent should log an *added* on the new subevent and an *deleted* on the old one.

File: src/pretix/base/models/items.py [hunk 10]
```diff
@@ -897,3 +897,31 @@ def _position_lookup(self) -> Q:
...
+        for variation in variations:
+            if variation.item not in items:
+                raise ValidationError(_('All variations must belong to an item contained in the items list'))
+                break
+
+    @staticmethod
+    def clean_items(event, items, variations):
+        for item in items:
+            if event != item.event:
+                raise ValidationError(_('One or more items does not belong to this event'))
```
Comment (raphaelm) [inline src/pretix/base/models/items.py]: plural items => "do not belong" instead of "does"

File: src/pretix/base/models/items.py [hunk 11]
```diff
@@ -897,3 +897,31 @@ def _position_lookup(self) -> Q:
 
     class QuotaExceededException(Exception):
         pass
+
+    @staticmethod
+    def clean_variations(items, variations):
+        for variation in variations:
+            if variation.item not in items:
+                raise ValidationError(_('All variations must belong to an item contained in the items list'))
```
Comment (raphaelm) [inline src/pretix/base/models/items.py]: Could you end the error messages with a full stop, if they are a full sentence? :)

File: src/pretix/base/models/items.py [hunk 12]
```diff
@@ -897,3 +897,31 @@ def _position_lookup(self) -> Q:
...
+                break
+
+    @staticmethod
+    def clean_items(event, items, variations):
+        for item in items:
+            if event != item.event:
+                raise ValidationError(_('One or more items does not belong to this event'))
+            if item.has_variations:
+                selected_vars = set(item.variations.all()).intersection(variations)
+                if bool(selected_vars) is False:
```
Comment (raphaelm) [inline src/pretix/base/models/items.py]: I find this hard to read and understand. Also, the query to item.variations.all() is not really required. How about:

``
if not any(var.item == item for var in variations):
``

File: doc/api/resources/quotas.rst (hunk 8 above)
Comment (turegjorup) [reply doc/api/resources/quotas.rst]: Correct - and body shouldn't contain `"id": 1,` - I'll have a word with the copy-paste monkey 🙈

File: src/pretix/base/models/items.py (hunk 12 above)
Comment (turegjorup) [reply src/pretix/base/models/items.py]: Much better - thanks. Still getting up to speed in python. Didn't know that construct.

File: src/pretix/api/views/item.py (hunk 9 above)
Comment (turegjorup) [reply src/pretix/api/views/item.py]: This should be addressed by commit [79c3579](https://github.com/pretix/pretix/pull/657/commits/79c3579db8e20a04ec3a49b874a8c1f399d353d7)

**Comment by turegjorup** [comment]
I think my last commits should address the remaining issues so I have removed the WIP label. 

If there you find additional problems please advice.
